File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestReference.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.Action;
import agi.foundation.compatibility.ArgumentException;
import agi.foundation.compatibility.AssertHelper;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.TypeLiteral;
import cesiumlanguagewriter.*;
import java.util.ArrayList;
import java.util.Collection;
import javax.annotation.Nonnull;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.runners.MethodSorters;
import org.junit.Test;

@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestReference {
    @Test
    public final void canConstructEscapedReferences() {
        String value = "identifier#property";
        Reference reference = new Reference(value);
        AssertHelper.assertEquals("identifier", reference.getIdentifier());
        final ArrayList<String> tempCollection$0 = new ArrayList<String>();
        tempCollection$0.add("property");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$0);
        AssertHelper.assertEquals(value, reference.getValue());
        value = "identifier#property.subProperty";
        reference = new Reference(value);
        AssertHelper.assertEquals("identifier", reference.getIdentifier());
        final ArrayList<String> tempCollection$1 = new ArrayList<String>();
        tempCollection$1.add("property");
        tempCollection$1.add("subProperty");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$1);
        AssertHelper.assertEquals(value, reference.getValue());
        value = "\\#identif\\\\\\\\\\#ier\\.\\\\#propertyName.\\.abc\\\\.def";
        reference = new Reference(value);
        AssertHelper.assertEquals("#identif\\\\#ier.\\", reference.getIdentifier());
        final ArrayList<String> tempCollection$2 = new ArrayList<String>();
        tempCollection$2.add("propertyName");
        tempCollection$2.add(".abc\\");
        tempCollection$2.add("def");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$2);
        AssertHelper.assertEquals(value, reference.getValue());
        value = "#propertyName.\\.abc\\\\.def";
        reference = new Reference(value);
        AssertHelper.assertIsEmpty(reference.getIdentifier());
        final ArrayList<String> tempCollection$3 = new ArrayList<String>();
        tempCollection$3.add("propertyName");
        tempCollection$3.add(".abc\\");
        tempCollection$3.add("def");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$3);
        AssertHelper.assertEquals(value, reference.getValue());
    }

    @Test
    public final void canConstructFromIdentifierAndProperty() {
        Reference reference = new Reference("identifier", "property");
        AssertHelper.assertEquals("identifier", reference.getIdentifier());
        final ArrayList<String> tempCollection$0 = new ArrayList<String>();
        tempCollection$0.add("property");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$0);
        AssertHelper.assertEquals("identifier#property", reference.getValue());
        reference = new Reference("#identif\\\\#ier.\\", "property.Name");
        AssertHelper.assertEquals("#identif\\\\#ier.\\", reference.getIdentifier());
        final ArrayList<String> tempCollection$1 = new ArrayList<String>();
        tempCollection$1.add("property.Name");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$1);
        AssertHelper.assertEquals("\\#identif\\\\\\\\\\#ier\\.\\\\#property\\.Name", reference.getValue());
    }

    @Test
    public final void canConstructFromIdentifierAndProperties() {
        final ArrayList<String> tempCollection$0 = new ArrayList<String>();
        tempCollection$0.add("property");
        tempCollection$0.add("subProperty");
        Reference reference = new Reference("identifier", tempCollection$0);
        AssertHelper.assertEquals("identifier", reference.getIdentifier());
        final ArrayList<String> tempCollection$1 = new ArrayList<String>();
        tempCollection$1.add("property");
        tempCollection$1.add("subProperty");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$1);
        AssertHelper.assertEquals("identifier#property.subProperty", reference.getValue());
        final ArrayList<String> tempCollection$2 = new ArrayList<String>();
        tempCollection$2.add("property.Name");
        tempCollection$2.add("subProperty");
        reference = new Reference("#identif\\\\#ier.\\", tempCollection$2);
        AssertHelper.assertEquals("#identif\\\\#ier.\\", reference.getIdentifier());
        final ArrayList<String> tempCollection$3 = new ArrayList<String>();
        tempCollection$3.add("property.Name");
        tempCollection$3.add("subProperty");
        AssertHelper.assertEquals(reference.getPropertyNames(), tempCollection$3);
        AssertHelper.assertEquals("\\#identif\\\\\\\\\\#ier\\.\\\\#property\\.Name.subProperty", reference.getValue());
    }

    @Test
    public final void throwsWithMissingDelimiter() {
        ArgumentException exception = AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            Reference unused = new Reference("MissingDelimiter");
        }));
        AssertHelper.assertStringContains("The provided reference string is not in the correct format", exception.getMessage());
    }

    @Test
    public final void throwsWithMissingDelimiterDueToEscaping() {
        ArgumentException exception = AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            Reference unused = new Reference("Missing\\#Delimiter");
        }));
        AssertHelper.assertStringContains("The provided reference string is not in the correct format", exception.getMessage());
    }

    @Test
    public final void throwsWithMissingProperties() {
        ArgumentException exception = AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            Reference unused = new Reference("MissingPropertyName#");
        }));
        AssertHelper.assertStringContains("The provided reference string is not in the correct format", exception.getMessage());
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }
}

File: DotNet/CesiumLanguageWriterTests/TestReference.cs
using System;
using System.Collections.Generic;
using CesiumLanguageWriter;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    [TestFixture]
    public class TestReference
    {
        [Test]
        public void CanConstructEscapedReferences()
        {
            string value = "identifier#property";
            var reference = new Reference(value);
            Assert.AreEqual("identifier", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property" });
            Assert.AreEqual(value, reference.Value);

            value = "identifier#property.subProperty";
            reference = new Reference(value);
            Assert.AreEqual("identifier", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property", "subProperty" });
            Assert.AreEqual(value, reference.Value);

            value = @"\#identif\\\\\#ier\.\\#propertyName.\.abc\\.def";
            reference = new Reference(value);
            Assert.AreEqual(@"#identif\\#ier.\", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "propertyName", @".abc\", "def" });
            Assert.AreEqual(value, reference.Value);

            value = @"#propertyName.\.abc\\.def";
            reference = new Reference(value);
            Assert.IsEmpty(reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "propertyName", @".abc\", "def" });
            Assert.AreEqual(value, reference.Value);
        }

        [Test]
        public void CanConstructFromIdentifierAndProperty()
        {
            var reference = new Reference("identifier", "property");
            Assert.AreEqual("identifier", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property" });
            Assert.AreEqual("identifier#property", reference.Value);

            reference = new Reference(@"#identif\\#ier.\", "property.Name");
            Assert.AreEqual(@"#identif\\#ier.\", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property.Name" });
            Assert.AreEqual(@"\#identif\\\\\#ier\.\\#property\.Name", reference.Value);
        }

        [Test]
        public void CanConstructFromIdentifierAndProperties()
        {
            var reference = new Reference("identifier", new List<string> { "property", "subProperty" });
            Assert.AreEqual("identifier", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property", "subProperty" });
            Assert.AreEqual("identifier#property.subProperty", reference.Value);

            reference = new Reference(@"#identif\\#ier.\", new List<string> { "property.Name", "subProperty" });
            Assert.AreEqual(@"#identif\\#ier.\", reference.Identifier);
            Assert.AreEqual(reference.PropertyNames, new List<string> { "property.Name", "subProperty" });
            Assert.AreEqual(@"\#identif\\\\\#ier\.\\#property\.Name.subProperty", reference.Value);
        }

        [Test]
        public void ThrowsWithMissingDelimiter()
        {
            var exception = Assert.Throws<ArgumentException>(() =>
            {
                var unused = new Reference("MissingDelimiter");
            });
            StringAssert.Contains("The provided reference string is not in the correct format", exception.Message);
        }

        [Test]
        public void ThrowsWithMissingDelimiterDueToEscaping()
        {
            var exception = Assert.Throws<ArgumentException>(() =>
            {
                var unused = new Reference(@"Missing\#Delimiter");
            });
            StringAssert.Contains("The provided reference string is not in the correct format", exception.Message);
        }

        [Test]
        public void ThrowsWithMissingProperties()
        {
            var exception = Assert.Throws<ArgumentException>(() =>
            {
                var unused = new Reference("MissingPropertyName#");
            });
            StringAssert.Contains("The provided reference string is not in the correct format", exception.Message);
        }
    }
}
